File: implementation/scipamato/scipamato-core-entity/src/test/java/ch/difty/scipamato/core/entity/CodeClassTest.java
package ch.difty.scipamato.core.entity;

import static ch.difty.scipamato.core.entity.CodeClass.CodeClassFields.DESCRIPTION;
import static ch.difty.scipamato.core.entity.CodeClass.CodeClassFields.NAME;
import static org.assertj.core.api.Assertions.assertThat;

import org.junit.jupiter.api.Test;

class CodeClassTest extends Jsr303ValidatedEntityTest<CodeClass> {

    private static final String JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE = "{javax.validation.constraints.NotNull.message}";
    private static final String DESC                                          = "this is cc1";

    CodeClassTest() {
        super(CodeClass.class);
    }

    @Override
    protected CodeClass newValidEntity() {
        return new CodeClass(1, "foo", "bar");
    }

    @Override
    protected String getToString() {
        return "CodeClass[id=1]";
    }

    @Override
    protected String getDisplayValue() {
        return newValidEntity().getId() + " - " + newValidEntity().getName();
    }

    @Test
    void validatingCodeClass_beingValid_succeeds() {
        verifySuccessfulValidation(new CodeClass(1, "foo", "bar"));
    }

    @Test
    void validatingCodeClass_withNullName_fails() {
        CodeClass cc = new CodeClass(1, null, "bar");
        validateAndAssertFailure(cc, NAME, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    void validatingCodeClass_withNullDescription_fails() {
        CodeClass cc = new CodeClass(1, "foo", null);
        validateAndAssertFailure(cc, DESCRIPTION, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE);
    }

    @Test
    void cloning_copiesValues() {
        CodeClass orig = new CodeClass(1, "cc1", DESC);
        CodeClass copy = new CodeClass(orig);
        assertThat(copy).isEqualToComparingFieldByField(orig);
    }

    @Test
    void sameValues_makeEquality() {
        CodeClass cc1 = new CodeClass(1, "cc1", DESC);
        CodeClass cc2 = new CodeClass(cc1);
        assertEquality(cc1, cc2);
    }

    private void assertEquality(CodeClass cc1, CodeClass cc2) {
        assertThat(cc1.equals(cc2)).isTrue();
        assertThat(cc2.equals(cc1)).isTrue();
        assertThat(cc1.hashCode()).isEqualTo(cc2.hashCode());
    }

    @Test
    void differingValues_makeInequality() {
        CodeClass cc1 = new CodeClass(1, "cc1", DESC);
        CodeClass cc2 = new CodeClass(2, "cc1", DESC);
        CodeClass cc3 = new CodeClass(1, "cc2", DESC);
        CodeClass cc4 = new CodeClass(1, "cc1", "this is cc2");

        assertThat(cc1.equals(cc2)).isFalse();
        assertThat(cc1.equals(cc3)).isFalse();
        assertThat(cc1.equals(cc4)).isFalse();
        assertThat(cc2.equals(cc3)).isFalse();
        assertThat(cc2.equals(cc4)).isFalse();
        assertThat(cc3.equals(cc4)).isFalse();

        assertThat(cc1.hashCode()).isNotEqualTo(cc2.hashCode());
        assertThat(cc1.hashCode()).isNotEqualTo(cc3.hashCode());
        assertThat(cc1.hashCode()).isNotEqualTo(cc4.hashCode());
        assertThat(cc2.hashCode()).isNotEqualTo(cc3.hashCode());
        assertThat(cc2.hashCode()).isNotEqualTo(cc4.hashCode());
        assertThat(cc3.hashCode()).isNotEqualTo(cc4.hashCode());
    }

    @SuppressWarnings({ "unlikely-arg-type", "EqualsWithItself", "ConstantConditions",
        "EqualsBetweenInconvertibleTypes" })
    @Test
    void equalingToSpecialCases() {
        CodeClass cc1 = new CodeClass(1, "cc1", DESC);

        assertThat(cc1.equals(cc1)).isTrue();
        assertThat(cc1.equals(null)).isFalse();
        assertThat(cc1.equals("")).isFalse();
    }

    @Test
    void differingValues_withIdNullOnOne() {
        CodeClass cc1 = new CodeClass(1, "cc1", DESC);
        CodeClass cc2 = new CodeClass(null, "cc1", DESC);
        assertInequality(cc1, cc2);
    }

    private void assertInequality(CodeClass cc1, CodeClass cc2) {
        assertThat(cc1.equals(cc2)).isFalse();
        assertThat(cc2.equals(cc1)).isFalse();
        assertThat(cc1.hashCode()).isNotEqualTo(cc2.hashCode());
    }

    @Test
    void differingValues_withIdNullOnBoth() {
        CodeClass cc1 = new CodeClass(null, "cc1", DESC);
        CodeClass cc2 = new CodeClass(null, "cc1", DESC);
        assertEquality(cc1, cc2);
    }

    @Test
    void differingValues_withNameNullOnOne() {
        CodeClass cc1 = new CodeClass(1, null, DESC);
        CodeClass cc2 = new CodeClass(1, "cc1", DESC);
        assertInequality(cc1, cc2);
    }

    @Test
    void differingValues_withNameNullOnBoth() {
        CodeClass cc1 = new CodeClass(1, null, DESC);
        CodeClass cc2 = new CodeClass(1, null, DESC);
        assertEquality(cc1, cc2);
    }

    @Test
    void differingValues_withDescriptionNullOnOne() {
        CodeClass cc1 = new CodeClass(1, "cc1", DESC);
        CodeClass cc2 = new CodeClass(1, "cc1", null);
        assertInequality(cc1, cc2);
    }

    @Test
    void differingValues_withDescriptionNullOnBoth() {
        CodeClass cc1 = new CodeClass(1, "cc1", null);
        CodeClass cc2 = new CodeClass(1, "cc1", null);
        assertEquality(cc1, cc2);
    }
}

File: implementation/scipamato/core/entity/src/test/kotlin/ch/difty/scipamato/core/entity/CodeClassTest.kt
package ch.difty.scipamato.core.entity

import ch.difty.scipamato.core.entity.CodeClass.CodeClassFields.DESCRIPTION
import ch.difty.scipamato.core.entity.CodeClass.CodeClassFields.NAME
import org.assertj.core.api.Assertions.assertThat

import org.junit.jupiter.api.Test

internal class CodeClassTest : Jsr303ValidatedEntityTest<CodeClass>(CodeClass::class.java) {

    override fun newValidEntity(): CodeClass {
        return CodeClass(1, "foo", "bar")
    }

    override val toString = "CodeClass[id=1]"
    override val displayValue = "${newValidEntity().id} - ${newValidEntity().name}"

    @Test
    fun validatingCodeClass_beingValid_succeeds() {
        verifySuccessfulValidation(CodeClass(1, "foo", "bar"))
    }

    @Test
    fun validatingCodeClass_withNullName_fails() {
        val cc = CodeClass(1, null, "bar")
        validateAndAssertFailure(cc, NAME, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE)
    }

    @Test
    fun validatingCodeClass_withNullDescription_fails() {
        val cc = CodeClass(1, "foo", null)
        validateAndAssertFailure(cc, DESCRIPTION, null, JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE)
    }

    @Test
    fun cloning_copiesValues() {
        val orig = CodeClass(1, "cc1", DESC)
        val copy = CodeClass(orig)
        assertThat(copy).isEqualToComparingFieldByField(orig)
    }

    @Test
    fun sameValues_makeEquality() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(cc1)
        assertEquality(cc1, cc2)
    }

    private fun assertEquality(cc1: CodeClass, cc2: CodeClass) {
        assertThat(cc1 == cc2).isTrue()
        assertThat(cc2 == cc1).isTrue()
        assertThat(cc1.hashCode()).isEqualTo(cc2.hashCode())
    }

    @Test
    fun differingValues_makeInequality() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(2, "cc1", DESC)
        val cc3 = CodeClass(1, "cc2", DESC)
        val cc4 = CodeClass(1, "cc1", "this is cc2")

        assertThat(cc1 == cc2).isFalse()
        assertThat(cc1 == cc3).isFalse()
        assertThat(cc1 == cc4).isFalse()
        assertThat(cc2 == cc3).isFalse()
        assertThat(cc2 == cc4).isFalse()
        assertThat(cc3 == cc4).isFalse()

        assertThat(cc1.hashCode()).isNotEqualTo(cc2.hashCode())
        assertThat(cc1.hashCode()).isNotEqualTo(cc3.hashCode())
        assertThat(cc1.hashCode()).isNotEqualTo(cc4.hashCode())
        assertThat(cc2.hashCode()).isNotEqualTo(cc3.hashCode())
        assertThat(cc2.hashCode()).isNotEqualTo(cc4.hashCode())
        assertThat(cc3.hashCode()).isNotEqualTo(cc4.hashCode())
    }

    @Test
    fun equalingToSpecialCases() {
        val cc1 = CodeClass(1, "cc1", DESC)

        assertThat(cc1 == cc1).isTrue()
        assertThat(cc1 == null).isFalse()
    }

    @Test
    fun differingValues_withIdNullOnOne() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(null, "cc1", DESC)
        assertInequality(cc1, cc2)
    }

    private fun assertInequality(cc1: CodeClass, cc2: CodeClass) {
        assertThat(cc1 == cc2).isFalse()
        assertThat(cc2 == cc1).isFalse()
        assertThat(cc1.hashCode()).isNotEqualTo(cc2.hashCode())
    }

    @Test
    fun differingValues_withIdNullOnBoth() {
        val cc1 = CodeClass(null, "cc1", DESC)
        val cc2 = CodeClass(null, "cc1", DESC)
        assertEquality(cc1, cc2)
    }

    @Test
    fun differingValues_withNameNullOnOne() {
        val cc1 = CodeClass(1, null, DESC)
        val cc2 = CodeClass(1, "cc1", DESC)
        assertInequality(cc1, cc2)
    }

    @Test
    fun differingValues_withNameNullOnBoth() {
        val cc1 = CodeClass(1, null, DESC)
        val cc2 = CodeClass(1, null, DESC)
        assertEquality(cc1, cc2)
    }

    @Test
    fun differingValues_withDescriptionNullOnOne() {
        val cc1 = CodeClass(1, "cc1", DESC)
        val cc2 = CodeClass(1, "cc1", null)
        assertInequality(cc1, cc2)
    }

    @Test
    fun differingValues_withDescriptionNullOnBoth() {
        val cc1 = CodeClass(1, "cc1", null)
        val cc2 = CodeClass(1, "cc1", null)
        assertEquality(cc1, cc2)
    }

    companion object {
        private val JAVAX_VALIDATION_CONSTRAINTS_NOT_NULL_MESSAGE = "{javax.validation.constraints.NotNull.message}"
        private val DESC = "this is cc1"
    }
}
